source/Engine.cpp

```diff
@@ -1444,152 +1453,165 @@ void Engine::CalculateStep()
-		else
-			for(const auto &sound : jumpSounds)
-				Audio::Play(sound.first);
-	}
-	// Check if the flagship just entered a new system.
-	if(flagship && playerSystem != flagship->GetSystem())
-	{
-		bool wormholeEntry = false;
-		// Wormhole travel: mark the wormhole "planet" as visited.
-		if(!wasHyperspacing)
-			for(const auto &it : playerSystem->Objects())
-				if(it.HasValidPlanet() && it.GetPlanet()->IsWormhole() &&
-						&it.GetPlanet()->GetWormhole()->WormholeDestination(*playerSystem) == flagship->GetSystem())
-				{
-					wormholeEntry = true;
-					player.Visit(*it.GetPlanet());
-				}
-
-		player.SetSystemEntry(wormholeEntry ? SystemEntry::WORMHOLE :
-			flagship->IsUsingJumpDrive() ? SystemEntry::JUMP :
-			SystemEntry::HYPERDRIVE);
-		doFlash = Preferences::Has("Show hyperspace flash");
-		playerSystem = flagship->GetSystem();
-		player.SetSystem(*playerSystem);
-		EnterSystem();
+		// Only process player commands and handle mouse clicks.
+		ai.MovePlayer(*player.Flagship(), activeCommands);
+		activeCommands.Clear();
+		HandleMouseClicks();
 	}
-	Prune(ships);
-
-	// Move the asteroids. This must be done before collision detection. Minables
-	// may create visuals or flotsam.
-	asteroids.Step(newVisuals, newFlotsam, step);
-
-	// Move the flotsam. This must happen after the ships move, because flotsam
-	// checks if any ship has picked it up.
-	for(const shared_ptr<Flotsam> &it : flotsam)
-		it->Move(newVisuals);
-	Prune(flotsam);
-
-	// Move the projectiles.
-	for(Projectile &projectile : projectiles)
-		projectile.Move(newVisuals, newProjectiles);
-	Prune(projectiles);
-
-	// Step the weather.
-	for(Weather &weather : activeWeather)
-		weather.Step(newVisuals, flagship ? flagship->Position() : center);
-	Prune(activeWeather);
-
-	// Move the visuals.
-	for(Visual &visual : visuals)
-		visual.Move();
-	Prune(visuals);
-
-	// Perform various minor actions.
-	SpawnFleets();
-	SpawnPersons();
-	GenerateWeather();
-	SendHails();
-	HandleMouseClicks();
-
-	// Now, take the new objects that were generated this step and splice them
-	// on to the ends of the respective lists of objects. These new objects will
-	// be drawn this step (and the projectiles will participate in collision
-	// detection) but they should not be moved, which is why we put off adding
-	// them to the lists until now.
-	ships.splice(ships.end(), newShips);
-	Append(projectiles, newProjectiles);
-	flotsam.splice(flotsam.end(), newFlotsam);
-	Append(visuals, newVisuals);
-
-	// Decrement the count of how long it's been since a ship last asked for help.
-	if(grudgeTime)
-		--grudgeTime;
-
-	// Populate the collision detection lookup sets.
-	FillCollisionSets();
-
-	// Perform collision detection.
-	for(Projectile &projectile : projectiles)
-		DoCollisions(projectile);
-	// Now that collision detection is done, clear the cache of ships with anti-
-	// missile systems ready to fire.
-	hasAntiMissile.clear();
+	else
+	{
```

Comment:
In my own code I would move all of this else clause into its own function, and just call that as e.g. `else StepGame();`

Reply:
Agreed. The big structural problem in this code which I didn't address is that input handling, simulation, and rendering are not cleanly separated. I stuck to doing a minimally invasive change, but ideally CalculateStep would be restructured.

How about putting the two branches of the if into respectively `CalculateTimeStep` and `CalculatePausedStep`?